internal/martian/proxy.go

```diff
@@ -806,26 +806,35 @@ func (p *Proxy) handle(ctx *Context, conn net.Conn, brw *bufio.ReadWriter) error
 		return p.handleUpgradeResponse(res, brw, conn)
 	}
 
-	if closing := p.writeResponse(res, brw, conn); closing {
+	if err := p.writeResponse(res, brw, conn); err != nil {
+		if errors.Is(err, errClose) {
+			return errClose
+		}
+
+		if isClosedConnError(err) {
+			log.Debugf(req.Context(), "connection closed prematurely while writing response: %v", err)
+		} else {
+			log.Errorf(req.Context(), "got error while writing response: %v", err)
+		}
 		return errClose
 	}
 
 	return nil
 }
 
-func (p *Proxy) writeResponse(res *http.Response, brw *bufio.ReadWriter, conn net.Conn) (closing bool) {
+func (p *Proxy) writeResponse(res *http.Response, brw *bufio.ReadWriter, conn net.Conn) error {
 	req := res.Request
 
 	if p.WriteTimeout > 0 {
 		if deadlineErr := conn.SetWriteDeadline(time.Now().Add(p.WriteTimeout)); deadlineErr != nil {
 			log.Errorf(req.Context(), "can't set write deadline: %v", deadlineErr)
 		}
+		defer conn.SetWriteDeadline(time.Time{})
```

Comment:
Nice catch!

Reply:
We should also reset the timer while reading.

Reply:
For reading it's ok due to tricky code.